Hard/WordBreakII.java

```diff
@@ -0,0 +1,57 @@
+import java.util.*;
+
+/**
+ * Word Break II
+ * Given a string s and a dictionary of words dict, add spaces in s to construct a sentence where each word is a valid dictionary word.
+ * Return all such possible sentences.
+ * For example, given
+ * s = "catsanddog",
+ * dict = ["cat", "cats", "and", "sand", "dog"].
+ * A solution is ["cats and dog", "cat sand dog"].
+ * Tags:  Dynamic Programming Backtracking
+ * @author chenshuna
+ */ 
+
+public class WordBreakII {
+    static HashMap<Integer, List<String>> map = new HashMap<>();
+
+    public static List<String> wordBreak(String s, Set<String> wordDict) {
+        int maxLengthOfWordDict = 0;
+        for(String str : wordDict)
+            maxLengthOfWordDict = Math.max(maxLengthOfWordDict, str.length());
+        return helper(s, wordDict, 0, maxLengthOfWordDict);
+    }
+
+    public static List<String> helper(String s, Set<String> wordDict, int start, int max){
```

Comment:
I can see you want to use max here to save a few iterations. I just doubt how much improvement it would have. For the worst case, you still need to iterate the whole string, right? Besides, it makes the logic more complicated. 

And you can make the logic simpler by breaking the word into just two parts 1) the prefix that is in dictionary 2) the suffix to further break. So you don't need `start`.